numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2018,6 +2049,119 @@ static PyObject *
     Py_INCREF(ret);
     return ret;
 }
+
+#if @is_half@
+static PyObject *
+half_hex(PyObject *self) {
+    npy_half half_val = PyArrayScalar_VAL(self, Half);
+
+    if (npy_half_isnan(half_val) || !npy_half_isfinite(half_val)) {
+        return halftype_str(self);
+    }
+
+    npy_float val = npy_half_to_float(PyArrayScalar_VAL(self, @Name@));
+
+    // xref: https://stackoverflow.com/a/15197157/5671364
+    char buf[
+        1 + // sign, '-' or '+'
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+        1 + // decimal point, '.'
+        1 + // mantissa-exponent separator, 'p'
+        1 + // mantissa sign, '-' or '+'
+        (NPY_SIZEOF_FLOAT * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+        1 // string terminator, '\0'
+    ];
+
+    npy_float_hex(val, buf);
+    float_to_half_hex(buf);
+    return PyUnicode_FromString(buf);
+}
+
+static PyObject *
+half_fromhex(PyObject *cls, PyObject *arg) {
+    const char *string, *string_end;
+    Py_ssize_t length;
+    PyObject *ret;
+    npy_float value;
+    npy_half half_value;
+
+    string = PyUnicode_AsUTF8AndSize(arg, &length);
+    if (string == NULL) {
+        return NULL;
+    }
+    string_end = string + length;
+
+    if (npy_float_fromhex(string, string_end, &value) == NPY_FALSE) {
+        return NULL;
+    }
+
+    ret = PyArrayScalar_New(Half);
+    if (ret == NULL) {
+        return NULL;
+    }
+
+    npy_clear_floatstatus_barrier((char *)&value);
+    half_value = npy_float_to_half(value);
+    if (npy_get_floatstatus_barrier((char *)&value) & NPY_FPE_OVERFLOW) {
```

Comment:
I just realized after pushing, if we go beyond float instead of beyond half, we'll return at line 2094's `if` and the message will be `.. too large for float` instead of `half`. Hmm, half is getting real tricky with a lot of edge cases. Will see how we can handle this.

Reply:
Did this get handled?

Reply:
I think wouldn't worry about this one.  Yes, it would be nicer to report half, but overall it doesn't matter too much.

Rather than float barrier, it may be a bit faster/easier to explicitly check for an `inf` result (and that it was not `inf` before) for scalar cases.

Reply:
That change probably fixes the piodide error indirectly:  It doesn't properly support FPEs.

Reply:
As a compromise, I removed the `@name@` so we don't get an incorrect message.